torch/distributed/device_mesh.py

```diff
@@ -488,6 +325,162 @@ def __init__(
+                - For world-size layouts with no overrides, it will try to use the default process group
+                - For other layouts, it creates appropriate subgroups using either split_group (more efficient and it only
+                available for nccl communicators) or new_group. The former sometimes is called "eager init" which means
+                we eagerly initialize the nccl communicator when we initialize the device mesh and the later is called
+                "lazy init" which means we lazily initialize the nccl communicator when we first hit first collective
+                - Because process group itself is not serializable, we only store the group name in the mapping
+            """
+            # Every mesh_dim_name will only map to one and only one layout.
+            # When the layout has backend initiated, we want to ensure that
+            root_mesh = _mesh_resources.get_root_mesh(self)
+            layouts_to_groups_map = _mesh_resources.layouts_to_groups.setdefault(
+                root_mesh, {}
+            )
+            names_to_layouts_map = _mesh_resources.names_to_layouts.setdefault(
+                root_mesh, {}
+            )
+            # Let mesh_dim_name maps to the current layout whose backend has not been initialized.
+            # If mesh_dim_name has already mapped to another layout, we throw an error.
+            if name is not None:
+                if (
+                    name in names_to_layouts_map
+                    and names_to_layouts_map[name] != layout
+                ):
+                    raise ValueError(
+                        f"Mesh_dim_name {name} has already mapped to layout {layout}."
+                    )
+                names_to_layouts_map[name] = layout
+
+            # If the layout has already been mapped to a process group backend, we ignore the
+            # backend_override and group, and return early. We also throw a warning for that.
+            if layout in layouts_to_groups_map:
+                # TODO: We need to bookkeep pg_option together with the group name.
+                if backend_override != (None, None):
+                    warnings.warn(
+                        f"Group for {layout} ({name=}) already exists, ignoring backend override"
+                    )
+                if group is not None:
+                    warnings.warn(
+                        f"Group for {layout} ({name=}) already exists, ignoring explicit group"
+                    )
+                return
+
+            # When user explicitly pass in a process group, we directly reuse that PG as backend rather
+            # than creating a new one.
+            if group is not None:
+                layouts_to_groups_map[layout] = group.group_name
+                return
+
+            default_group = _get_default_group()
+
+            if layout == _Layout(((get_world_size(), 1),)) and backend_override == (
+                None,
```

Comment:
you assume the default group has default backend settings here, but you don't know that.  if someone initialized the default backend unusually, but wants the default settings here, then what?

Reply:
an idea (for future PR) - maybe trying to infer whether the default_pg is a match is a bad idea. 
1) if users init the world group before creating a mesh, they can pass the default_group into the mesh init fn.  This is more explicit, and cleaner.  If they do not, we can just create a new world group.

This proposal would be less convenient but cleaner/explicit. we could discuss with folks if its better than the implicit but imprecise option we have currently.

Reply:
This is the same behavior as https://github.com/pytorch/pytorch/pull/159371/files#diff-dbbed99b01763453143e50565b636cb37f8f693aefaf18b57d621781114ed1b7.

Reply:
OK I record your second comment in a doc so that we don't lose tracking it.